topdown/crypto.go

```diff
@@ -126,6 +130,60 @@ func builtinCryptoX509ParseCertificateRequest(a ast.Value) (ast.Value, error) {
 	return ast.InterfaceToValue(x)
 }
 
+func builtinCryptoX509ParseRSAPrivateKey(a ast.Value) (ast.Value, error) {
+	// the input string is better to be base64 encoded in order to avoid the
+	// "space" or "new line" problem, because the pem.Decode does not like
+	// the "spaces" between lines in the key (it is expecting "new lines" instead)
+	input, err := builtins.StringOperand(a, 1)
+	if err != nil {
+		return nil, err
+	}
+
+	// string to bytes, needed for pem.decode as func arg
+	bytes := []byte(input)
+
+	// if the input is not a PEM string, attempt to decode b64
+	if str := string(input); !strings.HasPrefix(str, "-----BEGIN RSA PRIVATE KEY-----") {
+		bytes, err = base64.StdEncoding.DecodeString(str)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// decode the pem into the Block struct
+	p, _ := pem.Decode(bytes)
+	if p != nil && p.Type != blockTypeRSAPrivateKey {
+		return nil, fmt.Errorf("invalid PEM-encoded certificate signing request")
+	}
+	if p != nil {
+		// generally you want your key can hit this following statement,
+		// because it means your key is in a good shape (parse-able).
+		bytes = p.Bytes
+	}
+
```

Comment:
Can this block be simplified ? Something like:

```go
block, _ := pem.Decode([]byte(input))
if block == nil {
	return nil, fmt.Errorf("failed to parse PEM block containing the key")
}

priv, err := x509.ParsePKCS1PrivateKey(block.Bytes)
if err != nil {
	// handle error
}
```

Reply:
This part of the code are now split into 2 new functions (which is similar to the "getX509CertsFromString" and "getX509CertsFromPem" in the "builtinCryptoX509ParseAndVerifyCertificates")
- getRSAPrivateKeyFromString
- getRSAPrivateKeyFromPEM